### PR TITLE
Sanitation fix from Gogs

### PR DESCRIPTION
**Pull request by bkcsoft**

### CODE REVIEW

**Review by tboerger — APPROVED** [review]

**Comment by tboerger** [comment]
LGTM

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
LGTM

File: modules/markdown/sanitizer.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"regexp"
+	"sync"
+
+	"code.gitea.io/gitea/modules/log"
+
+	"github.com/microcosm-cc/bluemonday"
+
+	"github.com/gogits/gogs/modules/setting"
```
Comment (appleboy) [inline modules/markdown/sanitizer.go]: github.com/go-gitea/gitea

**Review by appleboy — CHANGES_REQUESTED** [review]

File: modules/markdown/sanitizer.go (hunk 1 above)
Comment (strk) [reply modules/markdown/sanitizer.go]: actually "code.gitea.io/gitea/modules/setting"

Comment (tboerger) [reply modules/markdown/sanitizer.go]: Good catch

Comment (lunny) [reply modules/markdown/sanitizer.go]: ^

File: modules/markdown/sanitizer_test.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Copyright 2017 The Gogs Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package markdown
+
+import (
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
```
Comment (lunny) [inline modules/markdown/sanitizer_test.go]: goconvey has been removed before. I think change this test file to `github.com/stretchr/testify` is better

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/markdown/sanitizer.go (hunk 1 above)
Comment (bkcsoft) [reply modules/markdown/sanitizer.go]: Noted

**Comment by bkcsoft** [comment]
I'm getting nil-panics in some tests 🙄

**Comment by bkcsoft** [comment]
And done! @appleboy @lunny please re-review 😄

**Comment by bkcsoft** [comment]
Now it's working!

File: modules/markdown/sanitizer.go [hunk 3]
```diff
@@ -48,10 +48,19 @@ func NewSanitizer() {
 
 // Sanitize takes a string that contains a HTML fragment or document and applies policy whitelist.
 func Sanitize(s string) string {
+	if sanitizer == nil {
```
Comment (lunny) [inline modules/markdown/sanitizer.go]: It seems a race problem?

File: modules/markdown/sanitizer.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	"regexp"
+	"sync"
+
+	"code.gitea.io/gitea/modules/log"
+
+	"github.com/microcosm-cc/bluemonday"
+
+	"code.gitea.io/gitea/modules/setting"
```
Comment (lunny) [inline modules/markdown/sanitizer.go]: merge the gitea internal packages.

File: modules/markdown/sanitizer.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		// Custom URL-Schemes
+		sanitizer.policy.AllowURLSchemes(setting.Markdown.CustomURLSchemes...)
+
+		log.Trace("Markdown: sanitizer initialized")
+	})
+}
+
+// Sanitize takes a string that contains a HTML fragment or document and applies policy whitelist.
+func Sanitize(s string) string {
+	if sanitizer.policy == nil {
```
Comment (lunny) [inline modules/markdown/sanitizer.go:50]: how about
```Go
func init() {
NewSanitizer(0
}
```
?

Comment (lunny) [reply modules/markdown/sanitizer.go:50]: But this maybe a race problem.

Comment (bkcsoft) [reply modules/markdown/sanitizer.go:50]: It uses `sync.Once` so there's no race-condition that I can see...

Comment (lunny) [reply modules/markdown/sanitizer.go:50]: OK

**Review by lunny — APPROVED** [review]

**Review by appleboy — APPROVED** [review]

**Comment by lunny** [comment]
let L-G-T-M work

**Comment by lunny** [comment]
@bkcsoft could you send a backport to v1.1.1?

**Pull request merged**